File: FishBun/src/main/java/com/sangcomz/fishbun/adapter/AlbumListAdapter.java
package com.sangcomz.fishbun.adapter;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.sangcomz.fishbun.R;
import com.sangcomz.fishbun.bean.Album;
import com.sangcomz.fishbun.define.Define;
import com.sangcomz.fishbun.ui.picker.PickerActivity;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.List;


public class AlbumListAdapter
        extends RecyclerView.Adapter<AlbumListAdapter.ViewHolder> {

    private List<Album> albumList;
    private ArrayList<Uri> pickedImagePath = new ArrayList<>();
    private int albumSize;
    private Bundle bundle;


    public AlbumListAdapter(ArrayList<Uri> pickedImagePath,
                            int albumSize,
                            Bundle bundle) {
        this.pickedImagePath = pickedImagePath;
        this.albumSize = albumSize;
        this.bundle = bundle;
    }

    public void setAlbumList(List<Album> albumList) {
        this.albumList = albumList;

    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.album_item, parent, false);
        return new ViewHolder(view, albumSize);
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, final int position) {
        holder.imgAlbumThumb.setImageDrawable(null);
        Picasso
                .with(holder.imgAlbumThumb.getContext())
                .load(Uri.parse(albumList.get(position).thumbnailPath))
                .fit()
                .centerCrop()
                .into(holder.imgAlbumThumb);
        holder.view.setTag(albumList.get(position));
        Album a = (Album) holder.view.getTag();
        holder.txtAlbumName.setText(albumList.get(position).bucketName);
        holder.txtAlbumCount.setText(String.valueOf(a.counter));


        holder.view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Album a = (Album) v.getTag();
                Context context = holder.view.getContext();
                Intent i = new Intent(context, PickerActivity.class);
                i.putExtra(Define.BUNDLE_NAME.ALBUM.name(), a);
                i.putExtra(Define.BUNDLE_NAME.POSITION.name(), position);
                i.putParcelableArrayListExtra(Define.INTENT_PATH, pickedImagePath);
                i.putExtras(bundle);
                ((Activity) context).startActivityForResult(i, new Define().ENTER_ALBUM_REQUEST_CODE);
            }
        });
    }

    @Override
    public int getItemCount() {
        return albumList.size();
    }

    public ArrayList<Uri> getPickedImagePath() {
        return pickedImagePath;
    }

    public void setPickedImagePath(ArrayList<Uri> pickedImagePath) {
        this.pickedImagePath = pickedImagePath;
    }

    public List<Album> getAlbumList() {
        return albumList;
    }


    public static class ViewHolder extends RecyclerView.ViewHolder {
        private View view;
        private ImageView imgAlbumThumb;
        private TextView txtAlbumName;
        private TextView txtAlbumCount;

        public ViewHolder(View view, int albumSize) {
            super(view);
            this.view = view;
            imgAlbumThumb = (ImageView) view.findViewById(R.id.img_album_thumb);
            imgAlbumThumb.setLayoutParams(new LinearLayout.LayoutParams(albumSize, albumSize));

            txtAlbumName = (TextView) view.findViewById(R.id.txt_album_name);
            txtAlbumCount = (TextView) view.findViewById(R.id.txt_album_count);
        }
    }
}

File: FishBun/src/main/java/com/sangcomz/fishbun/ui/album/adapter/AlbumListAdapter.kt
package com.sangcomz.fishbun.ui.album.adapter

import android.net.Uri
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView

import com.sangcomz.fishbun.R
import com.sangcomz.fishbun.adapter.image.ImageAdapter
import com.sangcomz.fishbun.ui.album.model.Album
import com.sangcomz.fishbun.ui.album.model.AlbumMetaData
import com.sangcomz.fishbun.ui.album.listener.AlbumClickListener
import com.sangcomz.fishbun.util.SquareImageView

class AlbumListAdapter(
    private val albumClickListener: AlbumClickListener,
    private val thumbnailSize: Int,
    private val imageAdapter: ImageAdapter
) :
    RecyclerView.Adapter<AlbumListAdapter.ViewHolder>() {

    private var albumList = emptyList<Album>()

    init {
        setHasStableIds(true)
    }

    override fun getItemId(position: Int): Long {
        return albumList[position].id
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(
            parent,
            thumbnailSize,
            imageAdapter
        ).apply {
            itemView.setOnClickListener {
                albumClickListener.onAlbumClick(adapterPosition, albumList[adapterPosition])
            }
        }
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.setData(albumList[position])
    }

    fun setAlbumList(albumList: List<Album>) {
        this.albumList = albumList
        notifyDataSetChanged()
    }

    fun updateAlbumMeta(position: Int, addedCount: Int, thumbnailPath: String) {
        val oldAlbum = albumList[position]
        val updateAlbum = oldAlbum.copy(
            metaData = AlbumMetaData(
                oldAlbum.metaData.count + addedCount,
                thumbnailPath
            )
        )
        albumList = albumList.toMutableList().apply { set(position, updateAlbum) }
        notifyItemChanged(position)
    }

    override fun getItemCount(): Int = albumList.size

    class ViewHolder(
        parent: ViewGroup,
        albumSize: Int,
        private val imageAdapter: ImageAdapter
    ) : RecyclerView.ViewHolder(
        LayoutInflater.from(parent.context).inflate(R.layout.album_item, parent, false)
    ) {
        private val imgAlbumThumb: SquareImageView = itemView.findViewById(R.id.img_album_thumb)
        private val txtAlbumName: TextView = itemView.findViewById(R.id.txt_album_name)
        private val txtAlbumCount: TextView = itemView.findViewById(R.id.txt_album_count)

        init {
            imgAlbumThumb.layoutParams = LinearLayout.LayoutParams(albumSize, albumSize)
        }

        fun setData(album: Album) {
            val uri: Uri = Uri.parse(album.metaData.thumbnailPath)
            imageAdapter.loadImage(imgAlbumThumb, uri)

            itemView.tag = album
            txtAlbumName.text = album.displayName
            txtAlbumCount.text = album.metaData.count.toString()
        }
    }
}